{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook showing a general workflow for the postprocessing analysis for the satay pipeline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os, sys\n",
    "import warnings\n",
    "import timeit\n",
    "import numpy as np\n",
    "import pandas as pd \n",
    "\n",
    "\n",
    "from transposonmapper.processing.clean_bedwigfiles import cleanfiles\n",
    "from transposonmapper.processing.transposonread_profileplot_genome import profile_genome\n",
    "from transposonmapper.processing.genomicfeatures_dataframe import dna_features\n",
    "from transposonmapper.statistics import volcano\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "wig_files=[]\n",
    "bed_files=[]\n",
    "pergene_files=[]\n",
    "#data_dir= \"../satay/data_files/data_unmerged/\"\n",
    "#data_dir=\"../transposonmapper/data_files/files4test/\"\n",
    "data_dir=\"../data/\"\n",
    "#data_dir=\"../transposonmapper/data_files/\"\n",
    "for root, dirs, files in os.walk(data_dir):\n",
    "    for file in files:\n",
    "        if file.endswith(\"sorted.bam.wig\"):\n",
    "            wig_files.append(os.path.join(root, file))\n",
    "        elif file.endswith(\"sorted.bam.bed\"):\n",
    "             bed_files.append(os.path.join(root, file))\n",
    "        elif file.endswith('sorted.bam_pergene_insertions.txt'):\n",
    "            pergene_files.append(os.path.join(root, file))\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir=\"../data/WT_1_Benoit\"\n",
    "wig_files=[\"../data/WT_1_Benoit/E-MTAB-4885.WT1.sorted.bam.wig\"]\n",
    "bed_files=[\"../data/WT_1_Benoit/E-MTAB-4885.WT1.sorted.bam.bed\"]\n",
    "pergene_files=[\"../data/WT_1_Benoit/ERR1533147_trimmed.sorted.bam_pergene_insertions.txt\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean the wig and bed files generated by transposon mapper (it is ok if run in spyder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wig file loaded ../data/WT_1_Benoit/E-MTAB-4885.WT1.sorted.bam.wig\n",
      "evaluating chromosome I\n",
      "evaluating chromosome II\n",
      "evaluating chromosome III\n",
      "evaluating chromosome IV\n",
      "evaluating chromosome V\n",
      "evaluating chromosome VI\n",
      "evaluating chromosome VII\n",
      "evaluating chromosome VIII\n",
      "evaluating chromosome IX\n",
      "evaluating chromosome X\n",
      "evaluating chromosome XI\n",
      "evaluating chromosome XII\n",
      "evaluating chromosome XIII\n",
      "evaluating chromosome XIV\n",
      "evaluating chromosome XV\n",
      "evaluating chromosome XVI\n",
      "Bed file loaded ../data/WT_1_Benoit/E-MTAB-4885.WT1.sorted.bam.bed\n",
      "evaluating chromosome I\n",
      "evaluating chromosome II\n",
      "evaluating chromosome III\n",
      "evaluating chromosome IV\n",
      "evaluating chromosome V\n",
      "evaluating chromosome VI\n",
      "evaluating chromosome VII\n",
      "evaluating chromosome VIII\n",
      "evaluating chromosome IX\n",
      "evaluating chromosome X\n",
      "evaluating chromosome XI\n",
      "evaluating chromosome XII\n",
      "evaluating chromosome XIII\n",
      "evaluating chromosome XIV\n",
      "evaluating chromosome XV\n",
      "evaluating chromosome XVI\n"
     ]
    }
   ],
   "source": [
    "## clean wig files for proper visualization in the genome Browser http://genome-euro.ucsc.edu/cgi-bin/hgGateway\n",
    "custom_header = \"\"\n",
    "split_chromosomes = False\n",
    "for files in zip(wig_files,bed_files):\n",
    "    cleanfiles(filepath=files[0], custom_header=custom_header, split_chromosomes=split_chromosomes)\n",
    "    cleanfiles(filepath=files[1], custom_header=custom_header, split_chromosomes=split_chromosomes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleanbed_files=[]\n",
    "for root, dirs, files in os.walk(data_dir):\n",
    "    for file in files:\n",
    "        if file.endswith(\"clean.bed\"):\n",
    "            cleanbed_files.append(os.path.join(root, file))\n",
    "\n",
    "cleanwig_files=[]\n",
    "for root, dirs, files in os.walk(data_dir):\n",
    "    for file in files:\n",
    "        if file.endswith(\"clean.wig\"):\n",
    "            cleanwig_files.append(os.path.join(root, file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome length:  12071326\n",
      "Reading file : /data/localhome/linigodelacruz/Documents/PhD_2018/Documentation/SATAY/src(source-code)/Transposonmapper/transposonmapper/data_files/Cerevisiae_AllEssentialGenes_List.txt\n",
      "Reading file : /data/localhome/linigodelacruz/Documents/PhD_2018/Documentation/SATAY/src(source-code)/Transposonmapper/transposonmapper/data_files/Cerevisiae_AllEssentialGenes_List.txt\n",
      "saving figure at ../data/WT_1_Benoit/E-MTAB-4885.WT1.sorted.bam_clean_transposonplot_genome.png\n"
     ]
    }
   ],
   "source": [
    "#transposonread_profileplot_genome.py (to check the insertion and read distribution throughout the genome)\n",
    "#example of file to analyse with profile_genome\n",
    "# bed_file=cleanbed_files[0]\n",
    "variable=\"transposons\" #\"reads\" \"transposons\"\n",
    "bar_width=None\n",
    "savefig=True\n",
    "\n",
    "for bed_file in cleanbed_files:\n",
    "\n",
    "    profile=profile_genome(bed_file=bed_file, variable=variable, bar_width=bar_width, savefig=savefig,showfig=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/WT_1_Benoit/E-MTAB-4885.WT1.sorted.bam_clean.wig']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleanwig_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selected region:  I\n",
      "Chromosome length =  230218\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  II\n",
      "Chromosome length =  813184\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  III\n",
      "Chromosome length =  316620\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  IV\n",
      "Chromosome length =  1531933\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  V\n",
      "Chromosome length =  576874\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  VI\n",
      "Chromosome length =  270161\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  VII\n",
      "Chromosome length =  1090940\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  VIII\n",
      "Chromosome length =  562643\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  IX\n",
      "Chromosome length =  439888\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  X\n",
      "Chromosome length =  745751\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  XI\n",
      "Chromosome length =  666816\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  XII\n",
      "Chromosome length =  1078177\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  XIII\n",
      "Chromosome length =  924431\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  XIV\n",
      "Chromosome length =  784333\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  XV\n",
      "Chromosome length =  1091291\n",
      "Everything alright, just ignore me!\n",
      "Selected region:  XVI\n",
      "Chromosome length =  948066\n",
      "Everything alright, just ignore me!\n"
     ]
    }
   ],
   "source": [
    "# genomic features \n",
    "# i=0\n",
    "\n",
    "# to make it only for certain files \n",
    "#cleanwig_files=[cleanwig_files[6],cleanwig_files[7],cleanwig_files[13]]\n",
    "for i in range(len(cleanwig_files)):\n",
    "    wig_file = cleanwig_files[i]\n",
    "    pergene_insertions_file = pergene_files[i]\n",
    "    plotting=False\n",
    "    variable=\"reads\" #\"reads\" or \"insertions\"\n",
    "    savefigure=False\n",
    "    verbose=True\n",
    "\n",
    "    dna_df2=[]\n",
    "    for chrom in ['I', 'II', 'III', 'IV','V', 'VI', 'VII', 'VIII', 'IX', 'X', 'XI', 'XII', 'XIII', 'XIV', 'XV', 'XVI']:\n",
    "    #     region=chrom\n",
    "    \n",
    "        region = chrom #e.g. 1, \"I\", [\"I\", 0, 10000\"], gene name (e.g. \"CDC42\")\n",
    "        dna_df2.append(dna_features(region=region,\n",
    "                    wig_file=wig_file,\n",
    "                    pergene_insertions_file=pergene_insertions_file,\n",
    "                    variable=variable,\n",
    "                    plotting=plotting,\n",
    "                    savefigure=savefigure,\n",
    "                    verbose=verbose))\n",
    "    data_genome=pd.concat(dna_df2, axis=0,ignore_index=True)\n",
    "    data_genome.to_excel('../postprocessed-data/'+ cleanwig_files[i].split(\"/\")[2] + \".xlsx\",engine='openpyxl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting wiggelen\n",
      "  Downloading wiggelen-0.4.1.tar.gz (22 kB)\n",
      "Building wheels for collected packages: wiggelen\n",
      "  Building wheel for wiggelen (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for wiggelen: filename=wiggelen-0.4.1-py3-none-any.whl size=25384 sha256=0cbff9853c8318585908900267233fb18aad932cd7bccc4e2a64b22bb407eeea\n",
      "  Stored in directory: /data/localhome/linigodelacruz/.cache/pip/wheels/5b/74/bd/712ac74845ac2f2ff60973943f18da1fc4718f2180ce2d0119\n",
      "Successfully built wiggelen\n",
      "Installing collected packages: wiggelen\n",
      "Successfully installed wiggelen-0.4.1\n"
     ]
    }
   ],
   "source": [
    "## Checking if the number of insertions that are in the wig file from Benoit is the same as the one he indicates in his paper\n",
    "\n",
    "!pip install wiggelen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import wiggelen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "ParseError",
     "evalue": "Could not parse line: variablestep chrom=chrI\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mParseError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-559db603bb91>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mwiggelen\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwalk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../data/WT_1_Benoit/E-MTAB-4885.WT1.sorted.bam_clean.wig'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'chr%s:%d\\t%s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/satay-dev/lib/python3.8/site-packages/wiggelen/wiggle.py\u001b[0m in \u001b[0;36mwalk\u001b[0;34m(track, force_index)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtrack\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 72\u001b[0;31m             \u001b[0mline_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstate\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     73\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mline_type\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mLineType\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mREGION\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m                 \u001b[0mregion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/satay-dev/lib/python3.8/site-packages/wiggelen/parse.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(line, state)\u001b[0m\n\u001b[1;32m     96\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mParseError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Could not parse line: %s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     97\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 98\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mParseError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Could not parse line: %s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mParseError\u001b[0m: Could not parse line: variablestep chrom=chrI\n"
     ]
    }
   ],
   "source": [
    "for x in wiggelen.walk(open('../data/WT_1_Benoit/E-MTAB-4885.WT1.sorted.bam_clean.wig')):\n",
    "    print('chr%s:%d\\t%s' % x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2730163c4b7859d47c676f1a1fe8f6449d0f2539972cd0f9489fa74976c27e38"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit ('satay-dev': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 2
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
